tiledb/sm/query/ast/query_ast.cc

```diff
@@ -45,6 +45,89 @@ static inline bool supported_string_type(Datatype type) {
       type == Datatype::STRING_UTF8);
 }
 
+ASTNodeVal::ASTNodeVal(
+    const std::string& field_name,
+    const void* const data,
+    uint64_t data_size,
+    const void* const offsets,
+    uint64_t offsets_size,
+    const QueryConditionOp op,
+    bool use_enumeration)
+    : field_name_(field_name)
+    , is_null_(false)
+    , op_(op)
+    , use_enumeration_(use_enumeration) {
+  if (data == nullptr) {
+    throw std::invalid_argument(
+        "ASTNodeVal set membership data must not be nullptr");
+  }
+
+  if (data_size == 0) {
+    throw std::invalid_argument(
+        "ASTNodeVal set membership data size must be greater than zero.");
+  }
+
+  if (offsets == nullptr) {
+    throw std::invalid_argument(
+        "ASTNodeVal set membership offsets must not be nullptr");
+  }
+
+  if (offsets_size == 0) {
+    throw std::invalid_argument(
+        "ASTNodeVal set membership offsets size must be greater than zero.");
+  }
+
+  if (offsets_size % sizeof(uint64_t) != 0) {
+    throw std::invalid_argument(
+        "ASTNodeVal set membership offsets is not a multiple of "
+        "uint64_t size.");
+  }
+
+  uint64_t num_offsets = offsets_size / 8;
+  auto offset_elems = static_cast<const uint64_t*>(offsets);
+  for (uint64_t i = 1; i < num_offsets; i++) {
+    if (offset_elems[i] < offset_elems[i - 1]) {
+      throw std::invalid_argument(
+          "ASTNodeVal set membership offsets must not decrease.");
+    }
+  }
+
+  if (offset_elems[num_offsets - 1] > data_size) {
+    throw std::invalid_argument(
+        "ASTNodeVal invalid set membership offsets invalid for data size: "
+        "last offset " +
+        std::to_string(offset_elems[num_offsets - 1]) +
+        " is larger than data size " + std::to_string(data_size));
+  }
+
+  if (op_ != QueryConditionOp::IN && op_ != QueryConditionOp::NOT_IN) {
+    throw std::invalid_argument(
+        "ASTNodeVal invalid set membership operator in set "
+        "membership constructor.");
+  }
+
+  data_ = ByteVecValue(data_size);
+  memcpy(data_.data(), data, data_size);
+
+  offsets_ = ByteVecValue(offsets_size);
+  memcpy(offsets_.data(), offsets, offsets_size);
+
+  for (uint64_t i = 0; i < num_offsets; i++) {
+    uint64_t start = offset_elems[i];
+    uint64_t length;
+
+    if (i + 1 < num_offsets) {
+      length = offset_elems[i + 1] - offset_elems[i];
+    } else {
+      length = data_size - offset_elems[i];
+    }
+
+    auto data_ptr = reinterpret_cast<char*>(data_.data() + start);
```

Comment:
Lets switch to static_cast<char*>(static_cast<void*>(data_.data() + start));

Reply:
Done.